File: mingle/app/models/plan/forecast.rb
#  Copyright 2020 ThoughtWorks, Inc.
#  
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License as
#  published by the Free Software Foundation, either version 3 of the
#  License, or (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
#  
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/agpl-3.0.txt>.

class Plan
  class Forecast
    HUNDRED_FIFTY_PERCENT_SCOPE_CHANGE = 2.5
    FIFTY_PERCENT_SCOPE_CHANGE = 1.5
    NO_SCOPE_CHANGE = 1.0

    class Base
      def to_json(options={})
        {
          :date => date,
          :no_velocity => no_velocity?,
          :late => late?,
          :scope => end_scope
        }.to_json(options)
      end
    end

    class Result < Base
      attr_reader :date, :deadline, :end_scope, :completed

      def initialize(forecasted_date, deadline, end_scope, completed)
        @date = forecasted_date
        @deadline = deadline
        @end_scope = end_scope
        @completed = completed
      end

      def no_velocity?
        false
      end

      def late?
        (@completed != @end_scope) && (@date && @date.beginning_of_day > @deadline.beginning_of_day)
      end
    end

    class NoVelocity < Base
      include Singleton
      def date; end
      def no_velocity?; true; end
      def late?; false; end
      def deadline; end
      def end_scope; end
    end

    extend Forwardable

    attr_reader :date

    def_delegator :@objective, :end_at, :end_date_of_objective

    def initialize(objective)
      @objective = objective
    end

    def for(project)
      snapshots = ObjectiveSnapshot.snapshots_till_date(@objective, project)
      
      velocity = Velocity.new(snapshots.map(&:completed), snapshots.map{|s| utc_time(s.dated).to_i })
      
      if velocity.invalid?
        {
          :name        => project.name.escape_html,
          :likely      => NoVelocity.instance,
          :less_likely => NoVelocity.instance,
          :not_likely  => NoVelocity.instance
        }
      else
        scope_values = snapshots.map(&:total)
        last_snapshot_date = utc_time(snapshots.last.dated)
        {
          :name        => project.name.escape_html,
          :likely      => convert(last_snapshot_date, *velocity.forecast(scope_values, HUNDRED_FIFTY_PERCENT_SCOPE_CHANGE)),
          :less_likely => convert(last_snapshot_date, *velocity.forecast(scope_values, FIFTY_PERCENT_SCOPE_CHANGE)),
          :not_likely  => convert(last_snapshot_date, *velocity.forecast(scope_values, NO_SCOPE_CHANGE))
        }
      end
    end

    def latest_date
      latest_completing_project = @objective.projects.max do |project1, project2|
        work_completion_date_for(project1) <=> work_completion_date_for(project2)
      end
      
      work_completion_date_for(latest_completing_project) if latest_completing_project
    end

    private
    def convert(last_snapshot_date, time_to_reach, end_scope, completed)
      Result.new(Time.at(last_snapshot_date.to_i + time_to_reach).utc.send(:to_date), end_date_of_objective, end_scope, completed)
    end
    
    def work_completion_date_for(project)
      self.for(project)[:not_likely].date || @objective.end_at
    end
    
    def utc_time(dated)
      ::Time.send("utc_time", dated.year, dated.month, dated.day)    
    end
  end
end

File: mingle/public/javascripts/planner/timeline/forecast.js
/*
*  Copyright 2020 ThoughtWorks, Inc.
*  
*  This program is free software: you can redistribute it and/or modify
*  it under the terms of the GNU Affero General Public License as
*  published by the Free Software Foundation, either version 3 of the
*  License, or (at your option) any later version.
*  
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU Affero General Public License for more details.
*  
*  You should have received a copy of the GNU Affero General Public License
*  along with this program.  If not, see <https://www.gnu.org/licenses/agpl-3.0.txt>.
*/
Timeline.Forecast = Class.create({
  initialize: function(objective, projectId, projectName) {
    this.objective = objective;
    this.projectId = projectId;
    this.projectName = projectName;
  },

  show: function(forecastData) {
    ObjectivesController.workProgress(this.objective.id, this.projectId, function(json) {
      var progress = json.progress;
      var startDate = this.parseDate(this.objective.start_at);
      var lastDate = this.parseDate(progress.last().date);
      var series = this.convertToSeries(forecastData, startDate, lastDate, progress);
      Timeline.Forecast.Chart.render(startDate, series);
    }.bind(this));
  },

  convertProgressDataToSeries: function(progress) {
    return progress.inject({actual: [], completed: []}, function(memo, snapshot) {
        memo.actual.push({x: this.parseDate(snapshot.date), y: snapshot.actual_scope});
        memo.completed.push({x: this.parseDate(snapshot.date), y: snapshot.completed_scope});
        return memo;
    }.bind(this));
  },

  convertToSeries: function(forecastData, startDate, lastDate, progress) {
    var progressSeries = this.convertProgressDataToSeries(progress);
    var notLikelyForecast = forecastData['not_likely'];
    var series = [
      this.series('Actual Scope', progressSeries.actual, "gray"),
      this.series('Completed Scope', progressSeries.completed, "#0973B6")
    ];
    if (notLikelyForecast.no_velocity) {
      return series;
    }

    var lessLikelyForecast = forecastData['less_likely'];
    var likelyForecast = forecastData['likely'];
    series.push({
      name: 'Projected end date',
      data: [
        [lastDate, progressSeries.completed.last().y],
        this.forecastInfo(notLikelyForecast),
        this.forecastInfo(lessLikelyForecast),
        this.forecastInfo(likelyForecast)
      ],
      dashStyle: 'dash',
      color: "#0973B6",
      lineWidth: 1,
      marker: {
          enabled: false,
          radius: 0,
          hover: {
              enabled: false
          }
      }
    });
    return series.concat([
      {startDate: startDate, forecast: lessLikelyForecast, mark: new Timeline.Forecast.Mark('50% scope change', 'gray', 'triangle')},
      {startDate: startDate, forecast: likelyForecast, mark: new Timeline.Forecast.Mark('150% scope change', 'gray', 'circle')},
      {startDate: lastDate, forecast: notLikelyForecast, mark: new Timeline.Forecast.Mark('No scope change', 'gray', 'square')}
    ].map(this.markSeries()));
  },

  markSeries: function() {
    return function(item) {
      var startX = item.startDate;
      var endX = this.parseDate(item.forecast.date);
      var y = item.forecast.scope;
      return item.mark.seriesData({x: startX, y: y}, {x: endX, y: y});
    }.bind(this);
  },

  series: function(seriesName, scopeSeries, color) {
    var series = new Timeline.Forecast.ContiguousSeries(seriesName, scopeSeries, color);
    return series.getData();
  },

  forecastInfo: function(forecast) {
    return [this.parseDate(forecast.date), forecast.scope];
  },

  parseDate: function(dateString) {
    var date = dateString.split("T")[0].split("-");
    return Date.UTC(parseInt(date[0], 10), parseInt(date[1] - 1, 10), parseInt(date[2], 10));
  }

});